### PR TITLE
Test for universal binaries macOS using Ninja

### LINKED ISSUE #17009: [bug] Generated CMake files for macOS universal binaries not compatible with Ninja
### Describe the bug

Using generated CMake files for macOS universal binaries aren't compatible with Ninja. For example:
```
# conanfile.txt
[requires]
zlib/1.3.1

[generators]
CMakeToolchain
CMakeDeps
```

```
# CMakeLists.txt
cmake_minimum_required(VERSION 3.18)
project(ninjatest CXX)

find_package(ZLIB CONFIG REQUIRED)
```

Please run the following commands:
```
$ conan install . -s arch="armv8|x86_64" -b missing -of build
$ cd build
$ cmake .. -GNinja \
    -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake \
    -DCMAKE_POLICY_DEFAULT_CMP0091=NEW \
    -DCMAKE_BUILD_TYPE=Release
```

Then you'll get something like
```
CMake Error:
  Running

   'ninja' '-C' '.../build' '-t' 'cleandead'

  failed with:

   ninja: error: build.ninja:81: expected newline, got '|'

```

The reason for this error is the phony target which refreshes the CMake cache
```
build ../CMakeLists.txt ... ZLIB-release-armv8|x86_64-data.cmake ...
```
which uses a filename with the non-escaped reserved symbol `|` in Ninja syntax.

Of course, one could argue that CMake should escape the referenced files in the generated `build.ninja` file correctly. On the other hand, it seems that Ninja doesn't support any escape sequence for the pipe character. So CMake got its hands tied in this matter. What is your take on this issue?

### How to reproduce it

_No response_

### PR DESCRIPTION
**Pull request by czoido**
Changelog: Fix: Replace `|` character in generated CMake and Environment files.
Docs: Omit

Closes: https://github.com/conan-io/conan/issues/17009

### CODE REVIEW

File: conan/tools/cmake/cmakedeps/templates/target_data.py [hunk 1]
```diff
@@ -23,7 +23,8 @@ def filename(self):
         if self.arch:
             data_fname += "-{}".format(self.arch)
         data_fname += "-data.cmake"
-        return data_fname
+        # https://github.com/conan-io/conan/issues/17009
+        return data_fname.replace("|", "-")
```
Comment (memsharded) [inline conan/tools/cmake/cmakedeps/templates/target_data.py]: Also in ``TargetConfigurationTemplate``

Comment (czoido) [reply conan/tools/cmake/cmakedeps/templates/target_data.py]: I checked and I think only the `build_type` ends up in that filename? 
https://github.com/conan-io/conan/blob/71795621a800c25021f27ab1fe73c5eefadad04e/conan/tools/cmake/cmakedeps/templates/target_configuration.py#L17
https://github.com/conan-io/conan/blob/71795621a800c25021f27ab1fe73c5eefadad04e/conan/tools/cmake/cmakedeps/cmakedeps.py#L26

**Comment by czoido** [comment]
We could also do the replaces to the `sh's` for the environment generated scripts.

File: conan/tools/cmake/cmakedeps/templates/target_data.py (hunk 1 above)
Comment (memsharded) [reply conan/tools/cmake/cmakedeps/templates/target_data.py]: Oh, yes, you are right!

File: test/functional/toolchains/cmake/test_universal_binaries.py [hunk 2]
```diff
@@ -99,3 +99,47 @@ def test(self):
...
+
+    client.run('install . -s=arch="armv8|x86_64" --build=missing -of=build')
+
+    with client.chdir("build"):
+        client.run_command("cmake .. -GNinja "
+                           "-DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake "
+                           "-DCMAKE_POLICY_DEFAULT_CMP0091=NEW "
+                           "-DCMAKE_BUILD_TYPE=Release")
+
+    assert "ninja: error: build.ninja:87: expected newline, got '|'" not in client.out
```
Comment (czoido) [inline test/functional/toolchains/cmake/test_universal_binaries.py]:
```suggestion
    assert "expected newline, got '|'" not in client.out
```

**Review by memsharded — APPROVED** [review]

**Pull request merged**